packages/salesforcedx-vscode-core/src/commands/forceSourceRetrieveMetadata/forceSourceRetrieveCmp.ts

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2019, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import {
+  Command,
+  SfdxCommandBuilder
+} from '@salesforce/salesforcedx-utils-vscode/out/src/cli';
+import {
+  ContinueResponse,
+  LocalComponent
+} from '@salesforce/salesforcedx-utils-vscode/out/src/types';
+import { RetrieveDescriber, RetrieveMetadataTrigger } from '.';
+import { nls } from '../../messages';
+import { TelemetryData } from '../../telemetry';
+import {
+  SfdxCommandlet,
+  SfdxCommandletExecutor,
+  SfdxWorkspaceChecker
+} from '../commands';
+import { RetrieveComponentOutputGatherer } from '../util/parameterGatherers';
+import { OverwriteComponentPrompt } from '../util/postconditionCheckers';
+
+export class ForceSourceRetrieveExecutor extends SfdxCommandletExecutor<
+  LocalComponent[]
+> {
+  private describer: RetrieveDescriber;
+
+  constructor(describer: RetrieveDescriber) {
+    super();
+    this.describer = describer;
+  }
+
+  public build(data?: LocalComponent[]): Command {
+    return new SfdxCommandBuilder()
+      .withDescription(nls.localize('force_source_retrieve_text'))
+      .withLogName('force_source_retrieve')
+      .withArg('force:source:retrieve')
+      .withArg('-m')
+      .withArg(this.describer.buildMetadataArg(data))
```

Comment:
Should we evaluate if `data` has a value before setting a command argument ?

Reply:
I have `data` as an optional property in the `buildMetadataArg` interface so this should be okay. That responsibility will fall to the implementation.